backend/backend-schema-proposal.md

```diff
@@ -39,59 +39,123 @@ GET `/request/{van_id}`
 GET `/location`
 
 - Subscribes the client to SSE triggered by new coordinates to any van
-- Responses are in the format `(lat, lon, route_id, van_id, timestamp)` where
-  `van_id` & `route_id` are 8-bit ints, `lat` & `lon` are 64-bit doubles, and
-  `timestamp` is a UNIX timestamp
 - Any response with an earlier timestamp than the most recent should be discarded
+- Excessive time drifting will be detected on backend and reported in admin panel
 - Any timestamp more than 1 min in the future should be discarded
 - Communication can be HTTP or HTTPS, but HTTPS is already being used
+- Response body is defined in JSON as per the following typescript spec:
 
-GET `/routes`
+``typescript
+interface Location {
+    vanId: number,
+    routeId: number,
+    location: Coordinate,
+    timestampMillis: number
+}
+``
 
-- Retrieves a predefined list of routes and their waypoints
+GET `/dashboard`
+
+- Retrieves all data to show on the main screen of the frontend
 - Coordinates are defined in a config file
-- Responses are a JSON array of routes in the following format:
+- Alerts with start times in the future or end times in the past are discarded
+- Response body is defined in JSON as per the following typescript spec:
 
 ```typescript
-[{
-    name: string
-    waypoints: [{
-        index: int,
-        stop: bool,
-        name: string | undefined,
-        longitude: double,
-        latitude: double,
-    }]
-}]
+interface Dashboard {
+    routes: Route[],
+    stops: Stop[],
+    alert: Alert | null
+}
+
+interface Alert {
+    body: string,
+    startTimestampSecs: number,
+    endTimestampSecs: number
+}
+
+interface Route {
+    id: number,
+    name: string,
+    active: boolean,
+    waypoints: Coordinate[] // Already ordered by index in backend
+}
+
+interface Stop {
+    id: number,
+    name: string,
+    active: boolean,
+    location: Coordinate
+}
 ```
 
 POST `/request`
 
 - Makes ADA request for the driver
 - The handler for this will decide which van is most appropriate and send an SSE
-- Request body is in the format `(lat, lon, timestamp)` where `lat` & `lon` are
-  64-bit doubles describing the client's coordinates
 - Any request with a timestamp more than 10 min out-of-date (or in the future)
```

Comment:
I kinda wish we were authenticating now, is there any way we can restrict the amount of ADA requests in a short time from a single user?

I know we can do something on the backend, but this seems like a major pain point if a malicious actor discovers this.

Reply:
I'd imagine FastAPI has some form of rate limiting built in.